src/charm.py

```diff
@@ -181,28 +173,30 @@ def _configure_amf(self, event: EventBase) -> None:  # noqa C901
                 return
         if not self._database_is_available():
             self.unit.status = WaitingStatus("Waiting for the amf database to be available")
-            event.defer()
             return
         if not self._get_database_info():
             self.unit.status = WaitingStatus("Waiting for AMF database info to be available")
-            event.defer()
             return
         if not self._nrf_requires.nrf_url:
             self.unit.status = WaitingStatus("Waiting for NRF data to be available")
-            event.defer()
             return
         if not self._amf_container.exists(path=CONFIG_DIR_PATH):
             self.unit.status = WaitingStatus("Waiting for storage to be attached")
-            event.defer()
             return
         if not _get_pod_ip():
             self.unit.status = WaitingStatus("Waiting for pod IP address to be available")
-            event.defer()
             return
+        if not self._private_key_is_stored():
+            self._generate_private_key()
         if not self._certificate_is_stored():
-            self.unit.status = WaitingStatus("Waiting for certificates to be stored")
-            event.defer()
-            return
+            if not self._csr_is_stored():
+                self._request_new_certificate()
+                self.unit.status = WaitingStatus("Waiting for certificates to be stored")
+                return
+            else:
+                provider_certificates = self._certificates.get_assigned_certificates()
+                if provider_certificates:
+                    self._store_certificate(certificate=provider_certificates[0].certificate)
```

Comment:
We should probably have another `if` before storing the certificate, there is no need to push the new cert if it is identical to the old one.

```python
if not self.certificate_is_stored(certificate):
    self._store_certificate(certificate)
```

We should also likely have a validation that the cert matches our CSR before storing it.

Reply:
Isn't that done by the library? https://github.com/canonical/self-signed-certificates-operator/blob/main/lib/charms/tls_certificates_interface/v3/tls_certificates.py#L1681

Reply:
If we have 2 csrs' in the relation databag (ex. if 1 cert has expired and we asked for a new one), `get_assigned_certificates` would return those 2 certificates in a list We would need to make sure that we use the right one out of those 2.

Reply:
Done